walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,163 @@
+      occupiedTimeRange.add(event.getWhen());
+    }
+  }
+
+  private boolean hasAttendeesAttending(Set<String> eventAttendees, Set<String> requestAttendees) {
+    for (String attendee: requestAttendees) {
+      if (eventAttendees.contains(attendee)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private List<TimeRange> getAvailableTimeRange(List<TimeRange> occupiedTimeRange, long duration) {
+    // Assumes 0 < duration <= WHOLE_DAY
+
+    List<TimeRange> availableMeetings = new ArrayList<>();
+
+    if (occupiedTimeRange.isEmpty()) {
+      availableMeetings.add(TimeRange.WHOLE_DAY);
+      return availableMeetings;
+    }
+
+    List<TimeRange> mergedTimeRanges = mergeTimeRange(occupiedTimeRange);
+
+    int start = TimeRange.START_OF_DAY;
+
+    for (TimeRange timeRange: mergedTimeRanges) {
+      // Interval duration is not enough.
+      if ((timeRange.start() - start) < duration) {
+        int startForInclusive = Math.min(timeRange.end() + 1, TimeRange.END_OF_DAY);
+        int startForNonInclusive = Math.min(timeRange.end(), TimeRange.END_OF_DAY);
+        start = timeRange.contains(timeRange.end()) ? startForInclusive : startForNonInclusive;
+        continue;
+      }
+
+      TimeRange availableMeeting = TimeRange.fromStartEnd(start, timeRange.start(), /* inclusive= */false);
+      availableMeetings.add(availableMeeting);
+
+      int startForInclusive = Math.min(timeRange.end() + 1, TimeRange.END_OF_DAY);
+      int startForNonInclusive = Math.min(timeRange.end(), TimeRange.END_OF_DAY);
+      start = timeRange.contains(timeRange.end()) ? startForInclusive : startForNonInclusive;
+    }
+
+    // Deals with the last remaining time slot of a day.
+    if (start < TimeRange.END_OF_DAY && (TimeRange.END_OF_DAY - start) >= duration) {
+      availableMeetings.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, /* inclusive= */true));
+    }
+
+    return availableMeetings;
+  }
+
+  private List<TimeRange> mergeTimeRange(List<TimeRange> timeRanges) {
+    List<TimeRange> mergedTimeRanges = new ArrayList<>();
+
+    if (timeRanges.isEmpty()) {
+      return mergedTimeRanges;
+    }
+
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+
+    TimeRange last = null;
+    for (TimeRange timeRange: timeRanges) {
+      // The first TimeRange or no overlapping.
+      if (last == null || !timeRange.overlaps(last)) {
+        mergedTimeRanges.add(timeRange);
+        last = timeRange;
+      } else {
+        TimeRange mergedTimeRange;
+        if (last.contains(timeRange)) {
+          continue;
+        }
+
+        if (last.end() == timeRange.end()) {
+          mergedTimeRange = TimeRange.fromStartEnd(last.start(), last.end(),
+                  last.contains(last.end()) || timeRange.contains(timeRange.end()));
+
+        } else {
+          mergedTimeRange = TimeRange.fromStartEnd(last.start(), timeRange.end(), timeRange.contains(timeRange.end()));
```

Comment:
as mentioned earlier, timeRange.contains(timeRange.end()) is always false

Reply:
.. my god, forget...fixed